File: lib/web/frontend/src/response_timer.ml
let get_cookie_value: string -> string = [%raw fun cookie_name -> {|
  var b = document.cookie.match('(^|;)\\s*' + cookie_name + '\\s*=\\s*([^;]+)');
  return b ? b.pop() : '';
|}]

let render_timer_text: (string -> string) -> string -> unit = [%raw fun get_text cookie_name -> {|
  var tag = document.getElementById('request-timer');
  tag.innerText = get_text(cookie_name);
|}]

let format_time time = match float_of_int (int_of_string time) with
  | t when t >= 1000000.0 -> (t /. 1000000.0, "s")
  | t when t >= 1000.0 -> (t /. 1000.0, "ms")
  | t -> (t, {js|µs|js})

let get_formatted cookie_name =
  let t, u = format_time (get_cookie_value cookie_name) in
    let rounded_t = match t with
      | r when float_of_int (int_of_float r) == t -> Js_float.toFixedWithPrecision t ~digits:0
      | _ -> Js_float.toFixedWithPrecision t ~digits:2 in
        {j| · Rendered in $rounded_t $u|j}

let set_timer_value () =
  render_timer_text get_formatted "mebe2-request-timer"

File: lib/web/frontend/src/response-timer.ts
import { LOCALE } from "./config.js";

const TIME_UNITS: Array<[number, string]> = [
  [10 ** 6, "s"],
  [10 ** 3, "ms"],
];

function getCookieValue(name: string): string {
  const b = document.cookie.match(`(^|;)\\s*${name}\\s*=\\s*([^;]+)`);
  return b && b.length > 0 ? b.pop() as string : "";
}

function formatTime(time: number): string {
  let formatUnit = "µs";
  for (const [limit, unit] of TIME_UNITS) {
    if (time >= limit) {
      time = time / limit;
      formatUnit = unit;
    }
  }

  const formatter = new Intl.NumberFormat(
    LOCALE,
    {
      style: "decimal",
      maximumFractionDigits: 2,
    },
  );

  return `${formatter.format(time)} ${formatUnit}`;
}

function getFormattedTime(cookieName: string): string {
  const time = parseInt(getCookieValue(cookieName), 10);

  if (!isNaN(time)) {
    return ` · ${formatTime(time)}`;
  } else {
    return "";
  }
}

export function renderTimeText() {
  const tag = document.getElementById("request-timer");

  if (tag === null) {
    throw new Error("Request timer DOM element missing!");
  }

  tag.innerText = getFormattedTime("mebe2-request-timer");
}
